src/solvers/nonlinear_solid_solver.cpp

```diff
@@ -0,0 +1,184 @@
+
+    mfem::GMRESSolver *J_gmres = new mfem::GMRESSolver(m_state[0].space->GetComm());
+    J_gmres->SetRelTol(m_lin_params.rel_tol);
+    J_gmres->SetAbsTol(m_lin_params.abs_tol);
+    J_gmres->SetMaxIter(m_lin_params.max_iter);
+    J_gmres->SetPrintLevel(m_lin_params.print_level);
+    J_gmres->SetPreconditioner(*m_J_prec);
+    m_J_solver = J_gmres;
+  } else {
+    mfem::HypreSmoother *J_hypreSmoother = new mfem::HypreSmoother;
+    J_hypreSmoother->SetType(mfem::HypreSmoother::l1Jacobi);
+    J_hypreSmoother->SetPositiveDiagonal(true);
+    m_J_prec = J_hypreSmoother;
+
+    mfem::MINRESSolver *J_minres = new mfem::MINRESSolver(m_state[0].space->GetComm());
+    J_minres->SetRelTol(m_lin_params.rel_tol);
+    J_minres->SetAbsTol(0.0);
+    J_minres->SetMaxIter(m_lin_params.max_iter);
+    J_minres->SetPrintLevel(m_lin_params.print_level);
+    J_minres->SetPreconditioner(*m_J_prec);
+    m_J_solver = J_minres;
+  }
+
+  if (m_timestepper == TimestepMethod::QuasiStatic) {
+    m_nonlinear_oper = new NonlinearSolidQuasiStaticOperator(m_H_form);
+  }
+  else {
+    // TODO: implement dynamic operator
+    mfem::mfem_error("Dynamic operator not implemented yet!");
+  }
+
+  // Set the newton solve parameters
+  m_newton_solver.iterative_mode = true;
+  m_newton_solver.SetSolver(*m_J_solver);
+  m_newton_solver.SetOperator(*m_nonlinear_oper);
+  m_newton_solver.SetPrintLevel(m_lin_params.print_level);
+  m_newton_solver.SetRelTol(m_lin_params.rel_tol);
+  m_newton_solver.SetAbsTol(m_lin_params.abs_tol);
+  m_newton_solver.SetMaxIter(m_lin_params.max_iter);
+
+}
+
+// Solve the Quasi-static Newton system
+void NonlinearSolidSolver::QuasiStaticSolve()
+{
+  mfem::Vector zero;
+  m_newton_solver.Mult(zero, *m_state[0].true_vec);
+}
+
+// Advance the timestep
+void NonlinearSolidSolver::AdvanceTimestep(__attribute__((unused)) double &dt)
+{
+  // Initialize the true vector
+  m_state[0].gf->GetTrueDofs(*m_state[0].true_vec);
+
+  if (m_timestepper == TimestepMethod::QuasiStatic) {
+    QuasiStaticSolve();
+  } else {
+    mfem::mfem_error("Only quasistatics implemented for nonlinear solid mechanics!");
+  }
+
+  // Distribute the shared DOFs
+  m_state[0].gf->SetFromTrueDofs(*m_state[0].true_vec);
+  m_cycle += 1;
+}
+
+NonlinearSolidSolver::~NonlinearSolidSolver() 
+{
+  delete m_H_form;
+  delete m_nonlinear_oper;
+  delete m_J_solver;
+  delete m_J_prec;
+  delete m_model;
+}
+
+NonlinearSolidQuasiStaticOperator::NonlinearSolidQuasiStaticOperator(mfem::ParNonlinearForm *H_form)
+  : mfem::Operator(H_form->FESpace()->GetTrueVSize())
+{
+  m_H_form = H_form;
+}
+
+// compute: y = H(x,p)
+void NonlinearSolidQuasiStaticOperator::Mult(const mfem::Vector &k, mfem::Vector &y) const
+{
+  // Apply the nonlinear form
+  m_H_form->Mult(k, y);
+}
+
+// Compute the Jacobian from the nonlinear form
+mfem::Operator &NonlinearSolidQuasiStaticOperator::GetGradient(const mfem::Vector &x) const
+{
+  m_Jacobian = &m_H_form->GetGradient(x);
```

Comment:
confusing (to me) operator precedence-- is this any different than just `return m_H_form->GetGradient(x)`?

Reply:
I don't think so. I save it in `m_Jacobian` because I can see a point where we might want to use it in other parts of the solver routines (for example returning sensitivities for the user). However, it may be better to just clean it up for the time being.

Reply:
Whoops! Good point, I thought it was a temporary at first glance.